Arrow 4139

mpeaton opened the pull request:
Introductory PR for ARROW-4139

scopatz commented:
Thanks @mpeaton! I think this should go to a topic branch, though, not master on the quansight fork

scopatz commented on python/pyarrow/parquet.py:
```diff
@@ -1407,3 +1406,51 @@ def read_schema(where, memory_map=False):
+
+    """
+    import pandas as pd
+    from numpy import datetime64
```
`np` is already imported at the top of the module

mpeaton replied:
Yes, thank you.  I did that as a hint toward specific importation, which I assumed is good practice.  I can remove it for the sake of efficiency.

scopatz commented on python/pyarrow/parquet.py:
```diff
@@ -1407,3 +1406,51 @@ def read_schema(where, memory_map=False):
+            lmin = datetime64(rg.column(i).statistics.min, 'us')
+            lmax = datetime64(rg.column(i).statistics.max, 'us')
+            df.loc[i] = [lmin, lmax, 'TIMESTAMP_MICROS']
+        else:
```
Is this the case for when it has a logical type of bytes (or byte_array)?

mpeaton replied:
line 1148 is a conditional ps.column(i).logical_type == 'TIMESTAMP_MICROS'
so, the highlighted code should not be executed for logical_type bytes.
I'm confused.  Can you expand upon your question?

pearu commented on python/pyarrow/tests/test_parquet.py:
```diff
@@ -2620,3 +2621,27 @@ def test_read_column_invalid_index():
+
+    table = pa.Table.from_pandas(df)
+
+    pq.write_table(table, 'example.parquet')
```
Use `tempdir` for creating temporary files, e.g. `pq.write_table(table, str(tempdir / 'example.parquet'))`

pearu reviewed (commented):
You should create the PR against upstream (https://github.com/apache/arrow.git), not Quansight/arrow master that should only track the upstream.

Changes to appveyor.yml, cpp/cmake_modules/ThirdpartyToolchain.cmake, cpp/src/arrow/CMakeLists.txt, cpp/src/arrow/flight/CMakeLists.txt, cpp/src/arrow/gpu/CMakeLists.txt, cpp/src/arrow/python/CMakeLists.txt, cpp/src/arrow/symbols.map, python/manylinux1/scripts/check_arrow_visibility.sh are not related to ARROW-4139 issue and should be reverted or proposed in a separate PR.

andygrove committed:
ARROW-5217: [Rust] [DataFusion] Fix failing tests

Due to recent changes in the HashMap implementation in Rust nightly, some non-deterministic tests started failing. This is a quick PR to fix this. It also seems that there are changes to `cargo fmt` so there are formatting changes in this PR too.

I commented out `csv_query_group_by_avg_multi_batch` for now as there is something more complex causing that one to fail. I can look at that over the weekend.

Author: Andy Grove <andygrove73@gmail.com>

Closes #4217 from andygrove/ARROW-5217 and squashes the following commits:

ad2bccaa <Andy Grove> fix build

mrkn committed:
ARROW-5150: [Ruby] Add Arrow::Table#raw_records

I'd like to add `raw_records` method in `Arrow::Table` class.
The same method of `RecordBatch` was added in #3587.

Author: Kenta Murata <mrkn@mrkn.jp>
Author: Kouhei Sutou <kou@clear-code.com>

Closes #4133 from mrkn/ruby_table_raw_records and squashes the following commits:

8d6b0d9c <Kouhei Sutou> Remove needless TODO
cc173459 <Kouhei Sutou> Unify tests for RecordBatch and Table
93967559 <Kenta Murata> Replace + with ++ for shifting row index
52cb4025 <Kenta Murata> Fix tags for errors
f7b88d4d <Kenta Murata> Add test cases for multiple columns and arrays
9ba352f7 <Kenta Murata> Add tests for StructArray
c47b5b34 <Kenta Murata> Add tests for ListArray
f6551fdc <Kenta Murata> Add tests for SparseUnionArray
a34c3603 <Kenta Murata> Add tests for DenseUnionArray
e8fb6c9b <Kenta Murata> Add Arrow::Table#raw_records
1e4f48ba <Kenta Murata> Rename

Commit:
ARROW-5226: [Gandiva] Add cmp functions for decimals

Author: Pindikura Ravindra <ravindra@dremio.com>

Closes #4219 from pravindra/ARROW-5226 and squashes the following commits:

bb1cb3dc <Pindikura Ravindra> ARROW-5226:  Add cmp functions for decimals

sbinet committed:
ARROW-5109: [Go] implement reading binary/string arrays from Arrow file

Author: Sebastien Binet <binet@cern.ch>

Closes #4155 from sbinet/issue-5109 and squashes the following commits:

f276be79 <Sebastien Binet> ARROW-5109:  implement reading binary/string arrays from Arrow file

sbinet committed:
ARROW-5111: [Go] implement reading list arrays from Arrow file

Author: Sebastien Binet <binet@cern.ch>

Closes #4157 from sbinet/issue-5111 and squashes the following commits:

a7fafdd9 <Sebastien Binet> ARROW-5111:  implement reading list arrays from Arrow file

sbinet committed:
ARROW-5110: [Go] implement reading struct arrays from Arrow file

Author: Sebastien Binet <binet@cern.ch>

Closes #4158 from sbinet/issue-5110 and squashes the following commits:

f6f9b9ea <Sebastien Binet> arrow/array: provide List array stringer implementation
b838ee04 <Sebastien Binet> arrow/array: fix Struct array Stringer implementation
d496c2ff <Sebastien Binet> ARROW-5110:  implement reading struct arrays from Arrow file

pitrou committed:
ARROW-4694: [CI] Improve detect-changes.py on Travis PRs

Author: Antoine Pitrou <antoine@python.org>

Closes #4223 from pitrou/ARROW-4694-improve-detect-changes and squashes the following commits:

d4cb46d2e <Antoine Pitrou> Move logic into get_travis_commit_range()
7b355e082 <Antoine Pitrou> Dump env vars
656eca4a8 <Antoine Pitrou> ARROW-4694:  Improve detect-changes.py on Travis PRs

Commit:
ARROW-5225: [Java] Improve performance of BaseValueVector#getValidityBufferSizeFromCount

Now in BaseValueVector#getValidityBufferSizeFromCount  and BitVectorHelper#getValidityBufferSize, it uses Math.ceil to calculate size which is not efficient (lots of unnecessary logic in StrictMath#floorOrCeil) . Since the valueCount is always not less than 0, we could simply replace Math.ceil with the following code:

return valueCount % 8 > 0 ? valueCount / 8 + 1 : valueCount / 8;

Author: tianchen <niki.lj@alibaba-inc.com>

Closes #4218 from tianchen92/ARROW-5225 and squashes the following commits:

e469c4a8 <tianchen> resolve comments
7f4a3a23 <tianchen> ARROW-5225: Improve performance of BaseValueVector#getValidityBufferSizeFromCount

Commit:
ARROW-5186 [Plasma] Fix crash caused by improper free on CUDA memory

Author: shengjun.li <shengjun.li@zilliz.com>

Closes #4177 from shengjun1985/master and squashes the following commits:

f384a9419 <shengjun.li> ARROW-5186  Fix crash caused by improper free on CUDA memory
48e0fda9e <shengjun.li> ARROW-5186  fix carsh on delete gpu memory

emkornfield committed:
ARROW-5240: [C++][CI] pin cmake_format

Hopefull this fixes the broken build.  Going to file a JIRA issue and file a ticket as this verifies a fix.

Author: Micah Kornfield <emkornfield@gmail.com>

Closes #4226 from emkornfield/pin_cmake_format and squashes the following commits:

30f6ef8e1 <Micah Kornfield> pin cmake_format

Commit:
ARROW-5243: [Java][Gandiva] Add decimal compare tests

Author: Pindikura Ravindra <ravindra@dremio.com>

Closes #4227 from pravindra/arrow-5243 and squashes the following commits:

4f0bd606 <Pindikura Ravindra> ARROW-5243:  Add decimal compare tests

eerhardt committed:
ARROW-4717 [C#] Consider exposing ValueTask instead of Task

Eliminating allocations in the C# Arrow library:

1. Use ValueTask instead of Task in the public APIs.
2. Eliminate allocations in ArrayData constructor by taking an array directly instead of calling `ToArray()` every time.
3. Remove some other minor allocations that provide no benefit.
4. Also, add a simple `NullCount == 0` check to optimize the case where there are no nulls in the data, but the null bitmap is filled with 0s instead of being empty.

With this change the `ArrowReaderWithMemory` benchmark in the repo goes from:

## Before
|                      Method |      Allocated Memory/Op |
|---------------------------- |--------------------:|
|       ArrowReaderWithMemory |  5.84 KB |

## After
|                      Method |      Allocated Memory/Op |
|---------------------------- |--------------------:|
|       ArrowReaderWithMemory |  4.48 KB |

@stephentoub @pgovind @chutchinson

Author: Eric Erhardt <eric.erhardt@microsoft.com>

Closes #4194 from eerhardt/ValueTask and squashes the following commits:

d2a00afd9 <Eric Erhardt> PR feedback.
310bb2f3b <Eric Erhardt> Convert all Task usages to ValueTask to save on allocations.
d1d2687d5 <Eric Erhardt> Minor optimizations

jorisvandenbossche committed:
ARROW-5237: [Python] populate _pandas_api.version

https://issues.apache.org/jira/browse/ARROW-5237

Author: Joris Van den Bossche <jorisvandenbossche@gmail.com>

Closes #4224 from jorisvandenbossche/pandas-metadata-version and squashes the following commits:

184a9736d <Joris Van den Bossche> ARROW-5237:  populate _pandas_api.version

sbinet committed:
ARROW-5172: [Go] implement reading fixed-size binary arrays from Arrow file

Author: Sebastien Binet <binet@cern.ch>

Closes #4159 from sbinet/issue-5172 and squashes the following commits:

6293fb0a <Sebastien Binet> arrow/array: use FixedWidthDataType interface
bba8571b <Sebastien Binet> ARROW-5172:  implement reading fixed-size binary arrays from Arrow file

sbinet committed:
ARROW-5173: [Go] handle multiple concatenated record batches

Author: Sebastien Binet <binet@cern.ch>

Closes #4160 from sbinet/issue-5173 and squashes the following commits:

623c3d2a <Sebastien Binet> ARROW-5173:  handle multiple concatenated record batches

sbinet committed:
ARROW-5174: [Go] implement Stringer for DataTypes

Author: Sebastien Binet <binet@cern.ch>

Closes #4161 from sbinet/issue-5174 and squashes the following commits:

54c92403 <Sebastien Binet> ARROW-5174:  implement Stringer for DataTypes

Commit:
ARROW-5232: [Java] Avoid runaway doubling of vector size

In reAlloc, double allocation size only if using the
current buffer capacity or the initial allocation as the base size.

Author: Pindikura Ravindra <ravindra@dremio.com>

Closes #4222 from pravindra/ARROW-5232 and squashes the following commits:

2a24673f <Pindikura Ravindra> ARROW-5232:  Avoid runaway doubling of vector size

mpeaton committed:
file based accessor for minmax statistics

mpeaton committed:
lint satisfaction
